src/api/index.js

```diff
@@ -0,0 +1,128 @@
+'use strict';
+
+import queryString from 'query-string';
+import isUrl from 'is-url';
+import axios from 'axios';
+import pkg from '../../package.json';
+
+/**
+ * @classdesc Represents an API call.
+ * @class
+ * @abstract
+ */
+export default class BynderApi {
+  /**
+   * Create a APICall.
+   * @constructor
+   * @param {String} baseURL A string with the base URL for account.
+   * @param {Object} httpsAgent A https agent.
+   * @param {Object} httpAgent A http agent.
+   * @param {String} token Optional OAuth2 access token
+   */
+  constructor(baseURL, httpsAgent, httpAgent, token) {
+    if (!isUrl(baseURL)) throw new Error('The base URL provided is not valid');
+
+    this.httpsAgent = httpsAgent;
+    this.httpAgent = httpAgent;
+    this.token = token;
+    this.axios = axios.create({ baseURL });
+  }
+
+  /**
+   * Generates the request headers
+   * @access private
+   * @async
+   * @param {String} method HTTP Verb of the request
+   * @param {Object} additionalHeaders Addional headers for specific endpoints
+   * @returns {Promise<object>} ?Request headers to be send
+   */
+  async _headers(method, additionalHeaders = {}) {
+    const headers = {
+      ...additionalHeaders,
+      'User-Agent': `bynder-js-sdk/${pkg.version}`
+    };
+
+    if (this.permanentToken) {
+      headers['Authorization'] = `Bearer ${this.permanentToken}`;
+    } else {
+      /* istanbul ignore next */
+      this.token = await (this.token.expired()
+        ? this.token.refresh()
+        : Promise.resolve(this.token));
+
+      headers['Authorization'] = `Bearer ${this.token.token.access_token}`;
+    }
+
+    return headers;
+  }
+
+  /**
+   * Fetch the information from the API.
+   * @async
+   * @param {String} method HTTP method
+   * @param {String} url URL to request
+   * @param {Object} params Body or querystring params to be sent on the request
+   * @param {Object} options Optional elements for the request
+   * @param {Object} options.additionalHeaders Additional headers specific to an endpoint
+   * @returns {Promise<object>} Object with response data or an Error with the problem.
+   */
+  async send(method, url, params = {}, options = {}) {
+
+    if (!this.token && !this.permanentToken) {
+      throw new Error('No token found');
+    }
+
+    const isV6orV7 = (/v[6|7]/).test(url);
+    let body = null;
+
+    if (method.toLocaleLowerCase() === 'post') {
+      body = params;
+
+      if (!isV6orV7) {
+        body = queryString.stringify(params);
+        // Some older endpoints require this
+        options.additionalHeaders = {
+          ...options.additionalHeaders,
+          'Content-Type': 'application/x-www-form-urlencoded'
+        };
+      }
```

Comment:
I'm not quite sure if this is actually the case, that 'Content-Type': 'application/x-www-form-urlencoded' for all "old" API endpoints.

And even if it's the case, I'd rather do such a refactor in a separate pull request.

Reply:
This in-line with the current release, which has it for all old endpoints. That's why I'm making the distinction between new and old endpoints.

However, I'm all in favour of moving this later to specific endpoints that actually required it.

Reply:
Well if we're sure this is the case for every `/api/` or `/v4/api/` request it's fine. Although in that case I think it would be a bit safer to check if the path starts with that, instead of checking if it starts with `/v6/` or `/v7/`.

Reply:
Fixed!